### PR TITLE
Support `href`/`text` element filters

### LINKED ISSUE #5598: Filtering element `href` by regex does not work
Trying to add an href filter containing URL path (/admin/sites/*/courses/42400-example-course), but it’s returning 0 results, when I use a wildcard (% or *), but when I search a specific path (/admin/sites/56306/courses/42400-example-course), it returns results. What am I doing wrong?

### PR DESCRIPTION
**Pull request by Twixes**
## Changes

Resolves #5598.

## Checklist

- [x] All querysets/queries filter by Organization, by Team, and by User
- [x] Django backend tests

### CODE REVIEW

File: ee/clickhouse/models/property.py [hunk 1]
```diff
@@ -225,50 +227,60 @@ def get_property_values_for_key(key: str, team: Team, value: Optional[str] = Non
...
+            else:
+                vals = [vals] if isinstance(vals, str) else vals
+                if operator.endswith("icontains"):
+                    attributes[key] = [f".*{re.escape(text)}.*" for text in vals]
+                elif operator.endswith("regex"):
+                    attributes[key] = vals
+                else:
+                    attributes[key] = [re.escape(text) for text in vals]
+    is_negated = operator in NEGATED_OPERATORS
+    if attributes:
```
Comment (macobo) [inline ee/clickhouse/models/property.py:270]: Relying on python empty objects being falsy is a bit shaky. Better to be explicit here.

File: ee/clickhouse/models/property.py [hunk 2]
```diff
@@ -225,50 +227,60 @@ def get_property_values_for_key(key: str, team: Team, value: Optional[str] = Non
...
     attributes: Dict[str, List] = {}
-
     for key in ["href", "text"]:
-        vals = filters.get(key)
-        if filters.get(key):
-            attributes[key] = [re.escape(vals)] if isinstance(vals, str) else [re.escape(text) for text in filters[key]]
-
-    if len(attributes.keys()) > 0:
-        or_conditions = []
+        if vals := filters.get(key):
```
Comment (macobo) [inline ee/clickhouse/models/property.py:258]: Code style nit: IMO relying in None here is shaky and makes this hard to follow since vals is only used N lines down.

How about:

```
if key in filters
```

However looking at the code a bit more I think it could benefit from extracting a function.

```
if key in filters:
  attributes[key] = get_element_match(filters[key], operator or "exact")

def get_element_match(values: Union[str, List[str]], operator: OperatorType):
  # Do the if-else here
```

File: ee/clickhouse/models/property.py [hunk 3]
```diff
@@ -225,50 +227,60 @@ def get_property_values_for_key(key: str, team: Team, value: Optional[str] = Non
...
+                attributes[key] = [r".*"]
+            else:
+                vals = [vals] if isinstance(vals, str) else vals
+                if operator.endswith("icontains"):
+                    attributes[key] = [f".*{re.escape(text)}.*" for text in vals]
+                elif operator.endswith("regex"):
+                    attributes[key] = vals
+                else:
+                    attributes[key] = [re.escape(text) for text in vals]
+    is_negated = operator in NEGATED_OPERATORS
```
Comment (macobo) [inline ee/clickhouse/models/property.py:269]: Style nit: I'd inline this given it's used N lines down and breaks up the flow of reading this complicated function.

File: ee/clickhouse/models/property.py [hunk 4]
```diff
@@ -225,50 +227,60 @@ def get_property_values_for_key(key: str, team: Team, value: Optional[str] = Non
...
+    if attributes:
         for key, value_list in attributes.items():
             for idx, value in enumerate(value_list):
-                params["{}_{}_{}_attributes_regex".format(prepend, key, idx)] = ".*?({}).*?".format(
-                    ".*?".join(['{}="{}"'.format(key, value)])
-                )
-                or_conditions.append("match(elements_chain, %({}_{}_{}_attributes_regex)s)".format(prepend, key, idx))
-            if len(or_conditions) > 0:
-                conditions.append("(" + (" OR ".join(or_conditions)) + ")")
+                optional_flag = "(?i)" if operator.endswith("icontains") else ""
```
Comment (macobo) [inline ee/clickhouse/models/property.py:273]: This is out of sync with what we were doing in property.py.

We should use `ILIKE` here as well instead of a custom solution.

File: ee/clickhouse/models/property.py [hunk 5]
```diff
@@ -225,50 +227,60 @@ def get_property_values_for_key(key: str, team: Team, value: Optional[str] = Non
...
+                params[f"{prepend}_{key}_{idx}_attributes_regex"] = f'{optional_flag}({key}="{value}")'
+                conditions.append(f"match(elements_chain, %({prepend}_{key}_{idx}_attributes_regex)s)")
+            if len(conditions) > 1:
+                conditions.append(f"({' AND '.join(conditions)})")
+    if conditions:
+        final_conditions = f"{'NOT ' if is_negated else ''}({' OR '.join(conditions)})"
+    else:
+        final_conditions = "" if is_negated else "0 = 192"
```
Comment (macobo) [inline ee/clickhouse/models/property.py:279]: This does not match the previous behavior - this is a new bug!

**Review by macobo — CHANGES_REQUESTED** [review]
The resulting code is hard to follow. Pointed out few suggestions.

Meta-feedback: Your PRs tend to scope-creep. In this case you also started "fixing" a bunch of logic around whether the function should return an empty filter vs `0 = 1` and how icontains works. It's made worse since these are implicit changes - not mentioned in the linked ticket, not mentioned in the PR itself.

These kind of changes make PRs painful to review, hard to verify whether all the cases are covered and make it more likely for bugs slip through the cracks. For the author it also makes it harder to get the darn thing green and in.

I'd suggest doing things in separate PRs - one for the bug fix and one for the behavioral change.

File: ee/clickhouse/models/property.py (hunk 5 above)
Comment (Twixes) [reply ee/clickhouse/models/property.py:279]: I don't think I see the bug, could you set out an example?

File: ee/clickhouse/models/property.py (hunk 4 above)
Comment (Twixes) [reply ee/clickhouse/models/property.py:273]: We already use regex for everything in here, I think this significantly simpler and less error-prone without an extra branch for `(not_)icontains`-only `ILIKE`.

File: ee/clickhouse/models/property.py (hunk 5 above)
Comment (macobo) [reply ee/clickhouse/models/property.py:279]: The behavior changed when calling `filter_element` from `format_action_filter` on an autocapture event. Having no conditions is completely valid there - previously this function returned an empty string, now it returns `0 = 1`

File: ee/clickhouse/models/property.py (hunk 4 above)
Comment (macobo) [reply ee/clickhouse/models/property.py:273]: See the main comment I made around this - I think this is scope creep and shouldn't be in this PR.


That said, a proper solution would still unify this with `properties.py` in terms of approaches to icontains. Having similar, but different approaches is just a recipe for bugs/inconsistencies down the line.

File: ee/clickhouse/models/property.py (hunk 1 above)
Comment (Twixes) [reply ee/clickhouse/models/property.py:270]: It's not shaky, it's the "pythonic" way of doing this.

Comment (Twixes) [reply ee/clickhouse/models/property.py:270]: It's not shaky, it's the "pythonic" way of doing this.

Comment (macobo) [reply ee/clickhouse/models/property.py:270]: Hmm, checked out [pep8](https://www.python.org/dev/peps/pep-0008/). Fair regarding sequences: 

`For sequences, (strings, lists, tuples), use the fact that empty sequences are false:`

This does not mention dicts but w/e. :) Might be a footgun in there if using your own collections class.

But keep this also in mind then:

```
Comparisons to singletons like None should always be done with is or is not, never the equality operators.

Also, beware of writing if x when you really mean if x is not None -- e.g. when testing whether a variable or argument that defaults to None was set to some other value. The other value might have a type (such as a container) that could be false in a boolean context!
```

We violate this just above in this file.

Comment (Twixes) [reply ee/clickhouse/models/property.py:298]: Yeah, the lacking null checks are a fair point, improved that.

File: ee/clickhouse/models/property.py (hunk 5 above)
Comment (Twixes) [reply ee/clickhouse/models/property.py]: Right – made things smarter.

**Comment by Twixes** [comment]
I agree the `icontains` fix that applies more generally (but is literally a single letter added in two places) could be a separate PR – made PR #5654 with it instead.
Disagree with this being scope creep besides that though – the problem here was that there was NO previous behavior – according to the UI these filters were perfectly possible, but in reality didn't do a thing. Hence I just added support for them to avoid any further confusion by users (like in the linked issue).

I added a bunch of tests here, so we have sensible coverage of the feature, and everything is green too.

File: ee/clickhouse/models/property.py (hunk 4 above)
Comment (Twixes) [reply ee/clickhouse/models/property.py]: Normal property filtering involves values from parsed JSON though, while in this case we use the elements directly, so the benefit of a regex is that i.e. we can handle double quote characters better.

File: ee/clickhouse/models/property.py (hunk 5 above)
Comment (macobo) [reply ee/clickhouse/models/property.py]: +1 good catch re behavior in models/property here :sweat_smile:

File: ee/clickhouse/models/property.py [hunk 6]
```diff
@@ -225,55 +229,94 @@ def get_property_values_for_key(key: str, team: Team, value: Optional[str] = Non
     )
 
 
-def filter_element(filters: Dict, prepend: str = "") -> Tuple[str, Dict]:
+def process_ok_values(ok_values: Any, operator: OperatorType) -> List[str]:
```
Comment (macobo) [inline ee/clickhouse/models/property.py:232]: Nittiest nit: I like placing functions like this at the end of a file - more important stuff at the top. This doesn't really matter though :sweat_smile:

**Review by macobo — APPROVED** [review]
Good work! Thanks for fixing this :)

File: ee/clickhouse/models/property.py (hunk 6 above)
Comment (Twixes) [reply ee/clickhouse/models/property.py]: I remember we talked about that at the offsite in October 😄 It's a fair point readability-wise
